test/common/stats/tag_extractor_impl_test.cc

```diff
@@ -31,6 +31,22 @@ TEST(TagExtractorTest, TwoSubexpressions) {
   EXPECT_EQ("cluster_name", tags.at(0).name_);
 }
 
+TEST(TagExtractorTest, IstioResponseCode) {
+  TagExtractorStdRegexImpl tag_extractor("response_code",
+                                         "(response_code=\\.=(.+?);\\.;)|_rq(_(\\d{3}))$");
```

Comment:
Does the non-capturing group `(?: ...)` not work for this use-case? I'm not sure what std::regex produces when you have this OR.

Reply:
this regex is copied from [here](https://github.com/istio/istio/blob/d7d55360475e1c0ef0150f006dddea7e5eeb3c62/tools/packaging/common/envoy_bootstrap.json#L48) with removing of `.`

istio use this from the beginning, but seems it's not working as expected?

Reply:
It depends on the regex engine I suppose: re2 vs std::regex. The issue is that there are three groups here, and it matters how they are numbered as well as how they are populated when optional:

1: outer removal groups
2: left option: response_code=.=
3. right option: _rq.

That means, to answer my question, non-capturing doesn't help. Does this issue happen with re2 or just regex? AFAIR re2 supports optional capturing groups, which can be utilized in envoy. I  hesitate to treat empty value as missing value (it might break for cases when the regex matches empty). 

@jmarantz for regex expertise.

Reply:
I think stats run on `StdRegex`, correct me if I'm wrong.

Reply:
The built-in stats run purely on RE2. However, for legacy reasons, we left the std::regex impl for API-driven tag extraction, despite the fact that it's significantly slower.

Maybe we should add a parallel path for RE2-based API-driven tag extraction.

Reply:
also, with respect to the capture-groups above, I'm not sure of the answer. I wouldn't say I'm a regex expert -- my knowledge is more around how to make systems run fast despite the presence of regexes, particularly the incredibly slow std::regex. I'll take a look at the PR.

Reply:
I didn't realize we cannot set RE2 for bootstrap tag extractors. That's probably a separate change, and I agree, we do need to support optional capturing groups in RE2, too.

Reply:
Can you make a parallel test for RE2? If there's a functional but for std::regex there might be one for re2 also.

Reply:
I'm working on trying to approach this by using 2 regex tag matchers rather than 1 with an OR.

But the regex in this testcase looks weird `(response_code=\\.=(.+?);\\.;)` . I really am not that good at regexes, and I don't understand what's with the 2 semi-colons, and why the response-code pattern doesn't looks like a 3-digit number; is that supposed to represent something other than an http response code?

Also there's nothing in this test that matches the first clause in the ORed regex, afaict.

Reply:
Also this specific case seems odd as a testcase.   Envoy's built-in regexes already includes:
```
addRe2(RESPONSE_CODE_CLASS, R"(_rq_((\d))xx$)", "_rq_");
```

and as it stands that cannot be overridden, but once I remove the dup-check you could just add in response_code=.... form.

Reply:
It kind of looks like `(response_code=\\.=(.+?);\\.;)` is trying to match on the prometheus formatted output, which doesn't make much sense. But the two `=` with any character in between doesn't quite fit.

Reply:
This is the Wasm artifact. It generates concatenated metrics like `key=.=value;.;` literally and then relies on tag extraction to get keys and values back. This is far from optimal, but it's difficult to create free-form metrics any other way in Envoy.

Reply:
This is why Envoy's API supports specifying tag name/value pairs explicitly during tag construction.

Would it make sense to iterate down that path? I think that might not be all it needs to for Wasm's use case but that could be changed, particularly if this is going to wind up sinking a lot of CPU encoding/decoding strings with regexes.

Reply:
In any case the first part of that ORed regex is redundant with what's built into Envoy.

Reply:
It would be great to stop creating mangled stat names, but we have looked twice there before, and I think the API doesn't fully support creating arbitrary key/value pairs explicitly as needed by Wasm. @douglas-reid I think was exploring it a few months ago.

Reply:
yes but that might be a better solution if we can get there, then going down the path of more mangling

What do you perceive is missing?

I think what's missing from my perspective is that if you create a stat "foo" with tags x:1, y:2, and another stat "foo" with x:1, y:3, I think it will not really work. The names would have to be uniquified. I didn't try to do this myself.

But what are you seeing as the missing link to utilize specifying tag name/value pairs during construction?

If that's what it is we could have that API append the tag names/values to the stat name so each instance is unique.

Reply:
@jmarantz iirc, the issue is that we don't have _all_ tag values at stats creation time. i believe the underlying envoy apis support dynamic tags, but that is not exposed (iirc) for wasm extensions. i agree that adding that capability to the Wasm ABI would be awesome (as would related features discussed in https://github.com/envoyproxy/envoy/issues/14071#issuecomment-1185084414).

Reply:
If you don't have all the tag-values at creation time, how can it work today? All the tag names and values must be encoded in the stat name when it is created.

The only question is whether we use a regex to extract tag names and values from the fully elaborated stat name, or whether we start with the names and values given explicitly,  and construct a unique tag name for them.

Reply:
The Wasm ABI just uses a string for defining a metric, see https://github.com/envoyproxy/envoy/blob/main/source/extensions/common/wasm/context.cc#L1247. So the key is to make it accept a list of tags and also do better about reusing symbols in those tags.

Reply:
I can try to take another stab at this issue. So my understanding is that we have to plug through tag list in several places:
1. Wasm ABI should take a list of tags, requires ABI bump (@PiotrSikora);
2. Envoy should preserve this list of tags for a custom metric.
3. Envoy needs to output the tags in prometheus-compatible format.

Reply:
This approach makes sense to me from a performance perspective, both because of the lack of the tag-list in the API, and the usage of a string_view. And there's another issue with the API (which I pointed out years ago when it was being built) -- that using a string_view for the name argument means you'll need to take a symbol-table lock every time a stat is created from wasm.

That's OK if it happens when a module is loaded, but should probably not be considered acceptable if running in the data-plane. Or has wasm evolved to where you can run some init code during loading to establish counters and gauges in wasm space that can be accessed by data threads afterward?


From a purely functional perspective, though, it should be OK to compose stat names that encode the tag names/values. But why not use the tag formats for `rq_XXX` that's already built into wasm? Why is the `response_code=XXX; ` needed as well?

Reply:
@jmarantz Re: locking. I don't think we can avoid adding a new key-value metric on the data path, that's just how the telemetry works in Istio mesh where we just don't know a-priori the clients or the servers we speak to during configuration. What we know is a list of key names, but not complete metrics. We are aware of the performance problems, and we don't require using this extension.

Re: format: The prometheus metric specification was ported from a different system so we have to keep the symbols the same.

Reply:
RE locking: OK you need to re-create the symbols in the data path; that's survivable. But you need to have an API that takes a StatName (or array of StatNames) so that they can be held on the wasm side (and maybe replicated per core) and passed into the API. Otherwise wasm stat creation will never scale to large numbers of cores.

RE "prometheus metric" OK I don't fully understand that but I'm OK moving forward with https://github.com/envoyproxy/envoy/pull/23024 if you think that helps. It's just that the test regex copied from this PR doesn't appear to make sense to me. We can just have the test case be a single alternate regex to populate `envoy.response_code`.

Reply:
I'm not deeply familiar with this Wasm code. Are you saying that Wasm context should hold on to storage better and be smarter about re-using already allocated symbols (per core)? Yes, I agree, and right now this is handled by the client code which avoids making ABI calls by caching those metric IDs.

Reply:
Leaving this to clients might be OK and it might not, depending on how long-lived the client's context is, typically. If it's a long time then the system will eventually settle into reasonable performance.

Consider a test case -- I'm not sure if this is realistic. Let's say there's 100 threads, and each of those has a separate wasm context. Now say there are 100k stats, each of which are comprised of different permutations of fixed keywords such as "upstream" or "http", names of specific clusters, which are quite common, and perhaps a dozen different response-code segments such as rq_200 and rq_2xx.

Now let's say (perhaps pessimistically) that the wasm module gets reloaded every minute for some reason related to the control plane.

This means the number of symbol table locks that will be needed before all the wasm-side caches are warm is 100 * 10k = 1M. I think there is a risk of a burst of contention every minute, which would affect long-tail latency.

We tried very hard to do in the stats API internal to C++, is to make it easy to save the common segments that comprise the stat names in pools or sets that are initialized once at startup. When a counter is created by concatenating together a bunch of pre-symbolized stat name segments, no locks are needed.

I think it would be beneficial to wasm long-tail performance to expose this API (StatNamePool, StatNameSet, createCounterFromStatNames, et al). But I also understand that my scenario above may be pessimistic in a few dimensions, and maybe in practice it's OK.

Reply:
Yes, our model does not handle xDS bursts very well. We use ECDS because we want to have exactly two copies of wasm contexts and the associated caches. I think the real reason why storage for us should be better exposed is because we have the opposite problem of metric expiry. We need to free symbols which I think doesn't happen because Wasm context uses one big arena for all symbols. Every new stat stays resident forever which has been noticed by some users that spin a lot of workloads with generated names.

Our team (and validated externally) opinion is that we should just rewrite our stats extension as a native extension, but we don't have available resources to do that.